src/middleware/express-error.js

```diff
@@ -0,0 +1,92 @@
+const uuid = require('uuid');
+const debug = require('debug')('api:errors');
+const ErrorReply = require('../error-reply');
+
+/**
+ * Create parameter validation middle-ware instance, given a mapping from
+ * parameter to regular expression or function that returns a message as string
+ * if the parameter is invalid.
+ *
+ * Parameters not listed in `req.params` will be ignored. But parameters
+ * present must match the pattern given in `options` or the request will be
+ * rejected with a 400 error message.
+ */
+const expressError = ({errorCodes, entry, monitor}) => {
+  const {name: method, cleanPayload} = entry;
+  return (err, req, res, next) => {
+
+    const incidentId = uuid.v4();
+    if (!(err instanceof ErrorReply)) {
+      err = new ErrorReply({
+        code: 'InternalServerError',
+        message: 'Internal Server Error, incidentId: ' + incidentId,
+        details: {incidentId}
+      });
+      debug(
+        'Error occurred handling: %s, err: %s, as JSON: %j, incidentId: %s',
+        req.url, err, err, incidentId, err.stack
+      );
+      if (monitor) {
+        err.incidentId = incidentId;
+        err.method = method;
+        err.params = req.params;
+        let payload = req.body;
+        if (cleanPayload) {
+          payload = cleanPayload(payload);
+        }
+        err.payload = req.payload;
+        monitor.reportError(err, {method});
```

Comment:
So I was about to mark this "approved", but I thought I'd just try it out testing a real application.  Happiliy, all of the tests passed

However, I added a `throw new Error('uhoh!');` to an API method in taskcluster-secrets -- just to test handling of internal errors -- and here's what was logged:
```
  api:errors Error occurred handling: /secret/captain%3AXkyBT0SfTW6kBYiyfwGLBw, err: Error: Internal Server Error, incidentId: 4cddbbce-d531-414e-89ce-8dfe8333ad60, as JSON: {"code":"InternalServerError","message":"Internal Server Error, incidentId: 4cddbbce-d531-414e-89ce-8dfe8333ad60","details":{"incidentId":"4cddbbce-d531-414e-89ce-8dfe8333ad60"}}, incidentId: 4cddbbce-d531-414e-89ce-8dfe8333ad60 Error: Internal Server Error, incidentId: 4cddbbce-d531-414e-89ce-8dfe8333ad60
    at /home/dustin/p/taskcluster-lib-api/src/middleware/express-error.js:20:13
    at Layer.handle_error (/home/dustin/p/taskcluster-lib-api/node_modules/express/lib/router/layer.js:71:5)
    at next (/home/dustin/p/taskcluster-lib-api/node_modules/express/lib/router/route.js:135:13)
    at Promise.resolve.then.then.catch (/home/dustin/p/taskcluster-lib-api/src/middleware/handle.js:35:14)
    at <anonymous>
    at runMicrotasksCallback (internal/process/next_tick.js:122:5)
    at _combinedTickCallback (internal/process/next_tick.js:132:7)
    at process._tickDomainCallback (internal/process/next_tick.js:219:9) +816ms
  taskcluster-lib-monitor reportError: Error: Internal Server Error, incidentId: 4cddbbce-d531-414e-89ce-8dfe8333ad60 level=[object Object] tags={} Error: Internal Server Error, incidentId: 4cddbbce-d531-414e-89ce-8dfe8333ad60
    at /home/dustin/p/taskcluster-lib-api/src/middleware/express-error.js:20:13
    at Layer.handle_error (/home/dustin/p/taskcluster-lib-api/node_modules/express/lib/router/layer.js:71:5)
    at next (/home/dustin/p/taskcluster-lib-api/node_modules/express/lib/router/route.js:135:13)
    at Promise.resolve.then.then.catch (/home/dustin/p/taskcluster-lib-api/src/middleware/handle.js:35:14)
    at <anonymous>
    at runMicrotasksCallback (internal/process/next_tick.js:122:5)
    at _combinedTickCallback (internal/process/next_tick.js:132:7)
    at process._tickDomainCallback (internal/process/next_tick.js:219:9) +814ms
```

So it's logging the "Internal Server Error" both with `debug`, the first section, and with `monitor.reportError`, the second section.  It's much more useful for debugging to log the *original* error, and only send the "Internal Server Error" to the API client.  Can you rearrange a little to make that work?

Reply:
Thanks for moving the `debug` call up.  I added a commit to also move the `monitor.reportError` call up (and add some comments).